Update Akka Persistence Cassandra

TimMoore opened the pull request:
## Fixes

Fixes #1195 

## Purpose

This is a branch to test against the latest version of Akka Persistence Cassandra (currently v0.89).

Versions from v0.80 include a schema change that requires a migration. We'd like to test compatibility and allow users to opt in to the new version in advance of making it the default for all Lagom Cassandra users.

- [x] bump versions
- [x] document (cross-link) migration
- [x] test the migration 
- [x] (extra) review all existing links to APC docs and `reference.conf` and point them to new [docs](https://doc.akka.io/docs/akka-persistence-cassandra/current/).

TimMoore commented:
Need to update to 0.89

TimMoore commented:
Updated to 0.89.

TimMoore commented:
Updated to 0.91

mikearnaldi commented:
Hi @TimMoore I have been using this for a while with lagom 1.4.x and never had any issue, I tried to update today to 1.5.0-RC1 and figured an issue, following a PR in a demo repo that shows the related changes applied to the master:
https://github.com/mikearnaldi/lagom-issue/pull/1/files

in the master I updated a basic app to use lagom 1.5.0-RC1 and tests are working fine.

The console log is huge but following the highlights of the error:
```
2019-02-15 14:29:49,928 ERROR com.lightbend.lagom.internal.persistence.testkit.AwaitPersistenceInit - Persistence failure when replaying events for persistenceId [persistenceInit5]. Last known sequence number [0]
scala.MatchError: OBJECT (of class com.typesafe.config.ConfigValueType)
        at akka.persistence.cassandra.package$.getListFromConfig(package.scala:140)
        at akka.persistence.cassandra.ConfigSessionProvider.lookupContactPoints(ConfigSessionProvider.scala:191)
        at akka.persistence.cassandra.ConfigSessionProvider.clusterBuilder(ConfigSessionProvider.scala:106)
        at akka.persistence.cassandra.ConfigSessionProvider.connect(ConfigSessionProvider.scala:44)
        at akka.persistence.cassandra.session.scaladsl.CassandraSession.setup$1(CassandraSession.scala:106)
        at akka.persistence.cassandra.session.scaladsl.CassandraSession.$anonfun$underlying$9(CassandraSession.scala:133)
        at akka.persistence.cassandra.session.scaladsl.CassandraSession.trySetup$1(CassandraSession.scala:159)
        at akka.persistence.cassandra.session.scaladsl.CassandraSession.retry(CassandraSession.scala:170)
        at akka.persistence.cassandra.session.scaladsl.CassandraSession.underlying(CassandraSession.scala:133)
        at akka.persistence.cassandra.session.scaladsl.CassandraSession.prepare(CassandraSession.scala:216)
        at akka.persistence.cassandra.snapshot.CassandraSnapshotStore.akka$persistence$cassandra$snapshot$CassandraSnapshotStore$$preparedSelectSnapshotMetadataWithMaxLoadAttemptsLimit(CassandraSnapshotStore.scala:74)
        at akka.persistence.cassandra.snapshot.CassandraSnapshotStore.loadAsync(CassandraSnapshotStore.scala:104)
        at akka.persistence.snapshot.SnapshotStore$$anonfun$receiveSnapshotStore$1.$anonfun$applyOrElse$1(SnapshotStore.scala:39)
        at akka.pattern.CircuitBreaker$State.materialize$1(CircuitBreaker.scala:718)
        at akka.pattern.CircuitBreaker$State.callThrough(CircuitBreaker.scala:755)
        at akka.pattern.CircuitBreaker$State.callThrough$(CircuitBreaker.scala:716)
        at akka.pattern.CircuitBreaker$Closed$.callThrough(CircuitBreaker.scala:827)
        at akka.pattern.CircuitBreaker$Closed$.invoke(CircuitBreaker.scala:836)
        at akka.pattern.CircuitBreaker.withCircuitBreaker(CircuitBreaker.scala:232)
        at akka.persistence.snapshot.SnapshotStore$$anonfun$receiveSnapshotStore$1.applyOrElse(SnapshotStore.scala:39)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
        at akka.actor.Actor.aroundReceive(Actor.scala:517)
        at akka.actor.Actor.aroundReceive$(Actor.scala:515)
        at akka.persistence.cassandra.snapshot.CassandraSnapshotStore.aroundReceive(CassandraSnapshotStore.scala:37)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)
        at akka.actor.ActorCell.invoke(ActorCell.scala:557)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
        at akka.dispatch.Mailbox.run(Mailbox.scala:225)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

TimMoore commented:
@mikearnaldi I think that's a bug in Akka Persistence Cassandra, introduced in 0.92 in https://github.com/akka/akka-persistence-cassandra/pull/417 and fixed in https://github.com/akka/akka-persistence-cassandra/pull/427 for the next upcoming release.

TimMoore commented:
The tests are all passing now, though it required increasing some test timeouts. I would say we should just accept that some of the tests are a bit slower, at least on Travis. They passed locally for me before increasing the timeouts.

Before merging, this will still need documentation updates.

renatocaval commented:
Thanks @TimMoore.
I will update the title to WIP.

 We are getting closer!! 🎉

ignasi35 commented on bin/scriptLib:
```diff
@@ -11,7 +11,7 @@ set -o pipefail
 export CURRENT_BRANCH=${TRAVIS_BRANCH}
 
-EXTRA_OPTS=""
+EXTRA_OPTS="-Dakka.test.timefactor=10"
```
Oh, we need to apply a `timefactor` to all tests? :-(

TimMoore replied:
I don't know if it really needs to be added to all tests, or if it needs to be increased by such a large factor, but I was experiencing several test failures specifically in Travis on this branch, and I thought it couldn't hurt. I could try to find a more optimized value if you think it's worth it, but it's a pretty time-consuming process to retest.

It looks like `master` hasn't had a passing build for a while, but the [build times on this branch](https://travis-ci.com/lagom/lagom/builds/114383197) seem roughly comparable to the times for passing jobs in recent builds:

* https://travis-ci.com/lagom/lagom/builds/114998394
* https://travis-ci.com/lagom/lagom/builds/114876996
* https://travis-ci.com/lagom/lagom/builds/114842956

So, I think this will only impact build times for legitimate failures.

TimMoore commented on persistence-cassandra/core/src/main/resources/play/reference-overrides.conf:
```diff
@@ -74,7 +74,7 @@ cassandra-snapshot-store {
 
 cassandra-query-journal {
-  first-time-bucket = "20160225"
+  first-time-bucket = "20160225T00:00"
```
Are you asking if the `timebucket` column needs to be changed in existing journal messages? Not that I'm aware of. @chbatey?

TimMoore commented on persistence-cassandra/core/src/main/resources/reference.conf:
```diff
@@ -172,6 +175,13 @@ lagom.persistence.read-side {
+
+    # Enable JMX reporter for Dropwizard Metrics.
+    # See: https://docs.datastax.com/en/developer/java-driver/3.5/manual/metrics/
+    jmx-reporting-enabled = on
```
This defaults to `off` for the journal plugin: https://github.com/akka/akka-persistence-cassandra/blob/3701c9c8e0ec12163f69293a5ac210d7cd28f50f/core/src/main/resources/reference.conf#L353

This group of settings are for the session created for read-side processors, which wouldn't read that setting. I only added settings that are used by `ConfigSessionProvider`.

TimMoore commented:
Updated to 0.99.

renatocaval commented:
I think this one is ready. We can add the documentation and migration guide on another PR

renatocaval commented:
This is ready (IMO) for a last review.

ignasi35 commented on docs/manual/java/guide/cluster/ReadSideCassandra.md:
```diff
@@ -102,5 +102,5 @@ Once you have finished registering all your event handlers, you can invoke the `
 The `CassandraSession` is using the [Datastax Java Driver for Apache Cassandra](https://github.com/datastax/java-driver).
 
-Each `ReadSideProcessor` instance is executed by an [Actor](https://doc.akka.io/docs/akka/2.5/actors.html?language=java) that is managed by [Akka Cluster Sharding](https://doc.akka.io/docs/akka/2.5/cluster-sharding.html?language=java). The processor consumes a stream of persistent events delivered by the `eventsByTag` [Persistence Query](https://doc.akka.io/docs/akka/2.5/persistence-query.html?la ...
+Each `ReadSideProcessor` instance is executed by an [Actor](https://doc.akka.io/docs/akka/2.5/actors.html?language=java) that is managed by [Akka Cluster Sharding](https://doc.akka.io/docs/akka/2.5/cluster-sharding.html?language=java). The processor consumes a stream of persistent events delivered by the `eventsByTag` [Persistence Query](https://doc.akka.io/docs/akka/2.5/persistence-query.html?la ...
```
```suggestion
Each `ReadSideProcessor` instance is executed by an [Actor](https://doc.akka.io/docs/akka/2.6/actors.html?language=java) that is managed by [Akka Cluster Sharding](https://doc.akka.io/docs/akka/2.6/cluster-sharding.html?language=java). The processor consumes a stream of persistent events delivered by the `eventsByTag` [Persistence Query](https://doc.akka.io/docs/akka/2.6/persistence-query.html?language=java) implemented by [akka-persistence-cassandra](https://doc.akka.io/docs/akka-persistence-cassandra/0.99/). The tag corresponds to the `tag` defined by the `AggregateEventTag`.
```


`s/2.5/2.6/`

ignasi35 replied:
(I mean, links to Akka docs use the segment `2.6` in the URL)

ignasi35 commented on docs/manual/java/releases/Migration16.md:
```diff
@@ -90,6 +90,14 @@ Switching from `persistence` to `ddata`, such as if your cluster relies of Lagom
 
+### Akka Persistence Cassandra Update
+
+The Akka Persistence Cassandra plugin is update to version 0.99. This version requires a schema migration before you upgrade to Lagom 1.6.0.
```
```suggestion
The Akka Persistence Cassandra plugin is updated to version 0.99. This version requires a schema migration before you upgrade to Lagom 1.6.0.
```

ignasi35 commented on docs/manual/java/releases/Migration16.md:
```diff
@@ -90,6 +90,14 @@ Switching from `persistence` to `ddata`, such as if your cluster relies of Lagom
+
+For more information on how to migrate, consult [Akka Persistence Cassandra migration document](https://doc.akka.io/docs/akka-persistence-cassandra/current/migrations.html#migrations-to-0-80-and-later).
+
+Note that although it's technically possible to run the migration while running your application we advice against it.
```
```suggestion
Note that although it's technically possible to run the migration while running your application we advise against it.
```

ignasi35 commented on docs/manual/java/releases/Migration16.md:
```diff
@@ -100,3 +108,4 @@ This is a summary of changes in Lagom 1.6 that would require a full cluster shut
 
 * The change in [[Akka Remote|Migration16#Remoting-Artery] default implementation.
 * The change in default [[Shard Coordination|Migration16#Shard-Coordination] strategy.
+* The change in [[Cassandra plugin version|Migration16#Akka-Persistence-Cassandra-Update]. Only impact Lagom applications using Cassandra.
```
```suggestion
* The change in [[Cassandra plugin version|Migration16#Akka-Persistence-Cassandra-Update]. Only impacts Lagom applications using Cassandra.
```

ignasi35 commented on docs/manual/scala/guide/cluster/PersistentEntityCassandra.md:
```diff
@@ -68,7 +68,7 @@ lagom.persistence.read-side.cassandra {
 With these properties set to `false`, if the keyspaces or tables are missing at startup your service will log an error and fail to start.
 
-Lagom's Cassandra support is provided by the [`akka-persistence-cassandra`](https://github.com/akka/akka-persistence-cassandra) plugin. A full configuration reference can be in the plugin's [`reference.conf`](https://github.com/akka/akka-persistence-cassandra/blob/v0.61/core/src/main/resources/reference.conf).
+Lagom's Cassandra support is provided by the [`akka-persistence-cassandra`](https://doc.akka.io/docs/akka-persistence-cassandra/0.99/) plugin. A full configuration reference can be in the plugin's [`reference.conf`](https://github.com/akka/akka-persistence-cassandra/blob/v0.99/core/src/main/resources/reference.conf).
```
```suggestion
Lagom's Cassandra support is provided by the [`akka-persistence-cassandra`](https://doc.akka.io/docs/akka-persistence-cassandra/0.99/) plugin. A full configuration reference is available in the plugin's [`reference.conf`](https://github.com/akka/akka-persistence-cassandra/blob/v0.99/core/src/main/resources/reference.conf).
```

ignasi35 commented on docs/manual/scala/guide/cluster/ReadSideCassandra.md:
```diff
@@ -102,5 +102,5 @@ Once you have finished registering all your event handlers, you can invoke the `
 The `CassandraSession` is using the [Datastax Java Driver for Apache Cassandra](https://github.com/datastax/java-driver).
 
-Each `ReadSideProcessor` instance is executed by an [Actor](https://doc.akka.io/docs/akka/2.5/actors.html?language=scala) that is managed by [Akka Cluster Sharding](https://doc.akka.io/docs/akka/2.5/cluster-sharding.html?language=scala). The processor consumes a stream of persistent events delivered by the `eventsByTag` [Persistence Query](https://doc.akka.io/docs/akka/2.5/persistence-query.html? ...
+Each `ReadSideProcessor` instance is executed by an [Actor](https://doc.akka.io/docs/akka/2.5/actors.html?language=scala) that is managed by [Akka Cluster Sharding](https://doc.akka.io/docs/akka/2.5/cluster-sharding.html?language=scala). The processor consumes a stream of persistent events delivered by the `eventsByTag` [Persistence Query](https://doc.akka.io/docs/akka/2.5/persistence-query.html? ...
```
```suggestion
Each `ReadSideProcessor` instance is executed by an [Actor](https://doc.akka.io/docs/akka/2.6/actors.html?language=scala) that is managed by [Akka Cluster Sharding](https://doc.akka.io/docs/akka/2.6/cluster-sharding.html?language=scala). The processor consumes a stream of persistent events delivered by the `eventsByTag` [Persistence Query](https://doc.akka.io/docs/akka/2.6/persistence-query.html?language=scala) implemented by [akka-persistence-cassandra](https://doc.akka.io/docs/akka-persistence-cassandra/0.99/). The tag corresponds to the `tag` defined by the `AggregateEventTag`.
```

ignasi35 commented on docs/manual/scala/releases/Migration16.md:
````diff
@@ -45,6 +45,13 @@ Switching from `persistence` to `ddata`, such as if your cluster relies of Lagom
 ```
+### Akka Persistence Cassandra Update
+
+The Akka Persistence Cassandra plugin is update to version 0.99. This version requires a schema migration before you upgrade to Lagom 1.6.0.
````
```suggestion
The Akka Persistence Cassandra plugin is updated to version 0.99. This version requires a schema migration before you upgrade to Lagom 1.6.0.
```

ignasi35 commented on docs/manual/scala/releases/Migration16.md:
```diff
@@ -45,6 +45,13 @@ Switching from `persistence` to `ddata`, such as if your cluster relies of Lagom
+
+For more information on how to migrate, consult [Akka Persistence Cassandra migration document](https://doc.akka.io/docs/akka-persistence-cassandra/current/migrations.html#migrations-to-0-80-and-later).
+
+Note that although it's technically possible to run the migration while running your application we advice against it.
```
```suggestion
Note that although it's technically possible to run the migration while running your application we advise against it.
```

ignasi35 commented on docs/manual/scala/releases/Migration16.md:
```diff
@@ -56,6 +63,7 @@ This is a summary of changes in Lagom 1.6 that would require a full cluster shut
 
 * The change in [[Akka Remote|Migration16#Remoting-Artery] default implementation.
 * The change in default [[Shard Coordination|Migration16#Shard-Coordination] strategy.
+* The change in [[Cassandra plugin version|Migration16#Akka-Persistence-Cassandra-Update]. Only impact Lagom applications using Cassandra.
```
```suggestion
* The change in [[Cassandra plugin version|Migration16#Akka-Persistence-Cassandra-Update]. Only impacts Lagom applications using Cassandra.
```

ignasi35 commented on docs/manual/java/guide/cluster/PersistentEntityCassandra.md:
```diff
@@ -68,7 +68,7 @@ lagom.persistence.read-side.cassandra {
 With these properties set to `false`, if the keyspaces or tables are missing at startup your service will log an error and fail to start.
 
-Lagom's Cassandra support is provided by the [`akka-persistence-cassandra`](https://github.com/akka/akka-persistence-cassandra) plugin. A full configuration reference can be in the plugin's [`reference.conf`](https://github.com/akka/akka-persistence-cassandra/blob/v0.61/core/src/main/resources/reference.conf).
+Lagom's Cassandra support is provided by the [`akka-persistence-cassandra`](https://doc.akka.io/docs/akka-persistence-cassandra/0.99/) plugin. A full configuration reference can be in the plugin's [`reference.conf`](https://github.com/akka/akka-persistence-cassandra/blob/v0.99/core/src/main/resources/reference.conf).
```
```suggestion
Lagom's Cassandra support is provided by the [`akka-persistence-cassandra`](https://doc.akka.io/docs/akka-persistence-cassandra/0.99/) plugin. A full configuration reference is available in the plugin's [`reference.conf`](https://github.com/akka/akka-persistence-cassandra/blob/v0.99/core/src/main/resources/reference.conf).
```

ignasi35 replied:
Nope, missed this one. :-)

TimMoore commented:
Thanks for getting this finished!
